src/anyp/Uri.cc

```diff
@@ -57,15 +57,28 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &encode)
     if (buf.isEmpty())
         return buf;
 
+    Parser::Tokenizer tk(buf);
+    const CharacterSet goodChars = encode.complement("ignore-chars");
```

Comment:
complement() is not an optimization. It is a requirement of this algorithm.

The encoder API is positive-form "encode these chars". The algorithm given is negative-form "do NOT encode UNLESS these chars".

Reply:
> complement() is not an optimization.

True. When you interpret my change request, please assume that I know about this fact.

> It is a requirement of this algorithm.

You can change the algorithm interface so that calling complement() is no longer an implementation requirement. That is my option 1. You do not _have_ to do that (there is option 2), but I do recommend option 1 for the reasons I have described in my change request.

> The encoder API is positive-form "encode these chars". The algorithm given is negative-form "do NOT encode UNLESS these chars".

Please feel free to ignore my comment that implies that calling complement is a "negative" action. IMO, the comment is correct, but it is not essential for justifying the requested changes.

FWIW, option 1 algorithm can be summarized as "given these good characters, encode all the bad ones" or "make sure the encoded version uses only the given characters". There are probably other, better ways to describe it. That description is of secondary importance here. The focus of this change request is on eliminating a prohibitively expensive operation from the encoding method. We can polish any associated wording later if needed.